design/one-pager-change-logs.md

```diff
@@ -0,0 +1,410 @@
+            volumeMounts:
+            - name: change-log-vol
+              mountPath: /var/run/change-logs
+          volumes:
+          - name: change-log-vol
+            emptyDir: {}
+```
+
+### Implementation Rollout
+
+The general sequence of implementation work to roll this feature out is
+described below.  As previously mentioned, the plan does not include any updates
+to core Crossplane to support this feature, as we will rely on a
+`DeploymentRuntimeConfig` to perform the injection of the sidecar container and
+the setup of a shared volume between the main provider and sidecar containers.
+The key code changes we will make are in `crossplane-runtime`, a new
+`change-logs-sidecar` repo, the Upjet framework and template, and the ecosystem
+of providers.
+
+* `crossplane-runtime`:
+  * Protobuf types are defined for change log entries and a gRPC client/server
+    to send these entries from client to server
+  * Managed reconciler is updated to generate change log entries, send them to
+  the gRPC server (i.e. sidecar container), and to include additional calls to
+  `Observe()` after each CUD operation to capture the "after" state of the entry
+* `change-logs-sidecar`
+  * A new repository is created to build and publish the image used in the sidecar container
+  * Consumes the protobuf types from `crossplane-runtime`
+  * Starts the gRPC server and listens for change log entries
+  * Writes all entries to `stdout` (i.e. pod logs)
+* [Upjet framework](https://github.com/crossplane/upjet):
+  * Controller templates are updated to use the new `crossplane-runtime` managed
+    reconciler logic and initialize it with a gRPC client
+* [Upjet provider
+  template](https://github.com/crossplane/upjet-provider-template):
+  * Consumes this new Upjet framework version
+  * [`main.go`](https://github.com/crossplane/upjet-provider-template/blob/main/cmd/provider/main.go)
+    is updated to create and initialize the gRPC client and then pass it into
+    the setup for all of its controllers
+* The major upjet based providers (e.g. `provider-upjet-aws`,
+  `provider-upjet-gcp`, `provider-upjet-azure`, etc.)
+  * Updated with changes similar to the upjet provider template (consumes new
+    Upjet version and creates gRPC client to pass into its controllers)
+
+### Assumptions
+
+Assumptions made for this proposal include:
+
+* Managed resources throughout the Crossplane ecosystem implement full
+  `spec.forProvider` and `status.atProvider` sections so that complete desired
+  and observed state is included in the change log entries simply by serializing
+  the full resource.
+
+## Future Work
+
+The below deliverables will not be included in the initial implementation of
+this feature, but should be considered in future milestones:
+
+* The [template repo](https://github.com/crossplane/provider-template) for
+  classic providers (not generated with Upjet) should be updated to include a
+  sidecar container, gRPC connection, etc.
```

Comment:
> should be updated to include a sidecar container

Do we really need any change in the provider codebase to include a sidecar container? Looks like this is done via `DeploymentRuntimeConfig`.

Reply:
Good catch! yep with the current approach of using a `DeploymentRuntimeConfig` the providers themselves do not need to be updated to set up the sidecar container. I'll update this.